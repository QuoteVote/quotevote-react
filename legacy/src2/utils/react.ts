import React, { 
  useState, 
  useEffect, 
  useRef, 
  ReactNode, 
  ChangeEvent, 
  KeyboardEvent,
  FormEvent,
  MouseEvent,
  Dispatch,
  SetStateAction
} from 'react';

export type { 
  ReactNode,
  ChangeEvent,
  KeyboardEvent,
  FormEvent,
  MouseEvent,
  Dispatch,
  SetStateAction
};

export { 
  React as default,
  useState,
  useEffect,
  useRef
};

// Add type declarations for common event handlers
export type InputChangeEvent = ChangeEvent<HTMLInputElement>;
export type TextAreaChangeEvent = ChangeEvent<HTMLTextAreaElement>;
export type ButtonClickEvent = MouseEvent<HTMLButtonElement>;
export type SetState<T> = Dispatch<SetStateAction<T>>;

// Add type declarations for common component props
export interface BaseProps {
  className?: string;
  style?: React.CSSProperties;
  children?: ReactNode;
}

export interface InputProps extends BaseProps {
  value: string;
  onChange: (value: string) => void;
  placeholder?: string;
  disabled?: boolean;
} 