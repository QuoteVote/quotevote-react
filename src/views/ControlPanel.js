import React from 'react'
import cx from 'classnames'
import Grid from '@material-ui/core/Grid'
import Typography from '@material-ui/core/Typography'
import Card from '@material-ui/core/Card'
import CardContent from '@material-ui/core/CardContent'
import Table from '@material-ui/core/Table'
import TableBody from '@material-ui/core/TableBody'
import TableCell from '@material-ui/core/TableCell'
import TableContainer from '@material-ui/core/TableContainer'
import TableHead from '@material-ui/core/TableHead'
import TableRow from '@material-ui/core/TableRow'
import Button from '@material-ui/core/Button'
import makeStyles from '@material-ui/core/styles/makeStyles'

const useStyles = makeStyles((theme) => ({
  panelContainer: {
    [theme.breakpoints.up('lg')]: {
      padding: 100,
      paddingTop: 30,
    },
    [theme.breakpoints.down('md')]: {
      padding: 10,
    },
    [theme.breakpoints.down('xs')]: {
      padding: 0,
    },
    backgroundColor: '#fafafa',
  },
  panelHeader: {
    font: 'Montserrat',
    fontSize: '24px',
    fontWeight: 'bold',
    color: 'black',
  },
  cardHeader: {
    font: 'Montserrat',
    fontSize: '18px',
    fontWeight: 'bold',
    letterSpacing: '0.2px',
    color: '#00cf6e',
  },
  sectionBorder: {
    height: '100%',
    borderRight: 'solid 2px #d2d2d2',
  },
  columnHeader: {
    font: 'Roboto',
    fontSize: '17px',
    color: '#00cf6e',
  },
  button: {
    width: '83.1px',
    height: '20.8px',
    borderRadius: '3px',
    textTransform: 'none',
    color: 'white',
  },
  pendingStatus: {
    borderRadius: '10px',
    padding: 2,
    backgroundColor: '#d8d8d8',
  },
  acceptedStatus: {
    borderRadius: '10px',
    padding: 2,
    backgroundColor: '#4caf50',
    color: 'white',
  },
  declinedStatus: {
    borderRadius: '10px',
    padding: 2,
    backgroundColor: '#da3849',
    color: 'white',
  },
}))

const ControlPanel = () => {
  const classes = useStyles()
  function createData(email, status) {
    return {
      email, status,
    }
  }
  const rows = [
    createData('testamail1@mail.test', 'Pending'),
    createData('testamail2@mail.test', 'Accepted'),
    createData('testamail3@mail.test', 'Declined'),
    createData('testamail4@mail.test', 'Pending'),
    createData('testamail5@mail.test', 'Pending'),
  ];
  const header = ['ID', 'Email', 'Status', 'Action']

  const renderActionButtons = (row) => {
    if (row.status === 'Pending') {
      return (
        <Grid container spacing={2} alignItems="center">
          <Grid item xs={6}>
            <Button
              variant="contained"
              color="secondary"
              className={classes.button}
              style={{
                backgroundColor: '#f44336',
              }}
            >
              Decline
            </Button>
          </Grid>
          <Grid item xs={6}>
            <Button
              variant="contained"
              color="primary"
              className={classes.button}
              style={{
                backgroundColor: '#00cf6e',
              }}
            >
              Accept
            </Button>
          </Grid>
        </Grid>
      )
    }

    return (
      <Grid container spacing={2} alignItems="center" justify="center">
        <Grid item xs={6}>
          <Button
            variant="contained"
            color="secondary"
            className={classes.button}
            style={{
              backgroundColor: '#f44336',
            }}
          >
            Decline
          </Button>
        </Grid>
      </Grid>
    )
  }
  return (
    <Grid container spacing={2} className={classes.panelContainer}>
      <Grid item xs={12}>
        <Typography className={classes.panelHeader}>Invite Control Panel</Typography>
      </Grid>
      <Grid container item xs={12}>
        <Grid container item xs={6} className={classes.sectionBorder}>
          <Grid item xs={11}>
            <Card>
              <CardContent>
                <Typography className={classes.cardHeader}>User Invitation Requests</Typography>
                <TableContainer>
                  <Table className={classes.table} aria-label="simple table">
                    <TableHead classes={{ head: classes.columnHeader }}>
                      <TableRow>
                        {header.map((name) => (
                          <TableCell align="center" className={classes.columnHeader}>{name}</TableCell>
                        ))}
                      </TableRow>
                    </TableHead>
                    <TableBody>
                      {rows.map((row, index) => (
                        <TableRow key={row.name}>
                          <TableCell align="center">
                            {index + 1}
                          </TableCell>
                          <TableCell align="center">{row.email}</TableCell>
                          <TableCell align="center">
                            <Typography
                              className={cx({
                                [classes.pendingStatus]: row.status === 'Pending',
                                [classes.declinedStatus]: row.status === 'Declined',
                                [classes.acceptedStatus]: row.status === 'Accepted',
                              })}
                            >
                              {row.status}
                            </Typography>
                          </TableCell>
                          <TableCell align="center">
                            {renderActionButtons(row)}
                          </TableCell>
                        </TableRow>
                      ))}
                    </TableBody>
                  </Table>
                </TableContainer>
              </CardContent>
            </Card>
          </Grid>
        </Grid>
        <Grid container item xs={6} justify="flex-end">
          <Grid item xs={11}>
            <Card>
              <CardContent>
                <Typography className={classes.cardHeader}>User Invitation Statistics</Typography>
              </CardContent>
            </Card>
          </Grid>
        </Grid>
      </Grid>
    </Grid>
  )
}

export default ControlPanel
