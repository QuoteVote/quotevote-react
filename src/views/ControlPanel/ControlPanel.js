import React from 'react'
import PropTypes from 'prop-types'
import cx from 'classnames'
import Grid from '@material-ui/core/Grid'
import Typography from '@material-ui/core/Typography'
import Card from '@material-ui/core/Card'
import CardContent from '@material-ui/core/CardContent'
import Table from '@material-ui/core/Table'
import TableBody from '@material-ui/core/TableBody'
import TableCell from '@material-ui/core/TableCell'
import TableContainer from '@material-ui/core/TableContainer'
import TableHead from '@material-ui/core/TableHead'
import TableRow from '@material-ui/core/TableRow'
import Button from '@material-ui/core/Button'
import Skeleton from '@material-ui/lab/Skeleton'
import makeStyles from '@material-ui/core/styles/makeStyles'

import { useMutation, useQuery } from '@apollo/react-hooks'
import { USER_INVITE_REQUESTS } from 'graphql/query'
import { UPDATE_USER_INVITE_STATUS } from 'graphql/mutations'

// react plugin for creating charts
import ChartistGraph from 'react-chartist'
import Chartist from 'chartist'
import { dailySalesChart } from 'variables/charts'
import moment from 'moment'
import controlPanelStylwa from './controlPanelStyles'

const useStyles = makeStyles(controlPanelStylwa)

const ActionButtons = ({ status, id }) => {
  const classes = useStyles()
  const [sendUserInviteApproval, { loading }] = useMutation(UPDATE_USER_INVITE_STATUS)
  const submitData = async (selectedStatus) => {
    await sendUserInviteApproval({
      variables: {
        userId: id,
        inviteStatus: `${selectedStatus}`,
      },
      refetchQueries: [{
        query: USER_INVITE_REQUESTS,
      }],
    })
  }

  const handleDecline = async () => {
    await submitData(2)
  }

  const handleAccept = async () => {
    await submitData(4)
  }

  const handleReset = async () => {
    await submitData(1)
  }

  switch (Number(status)) {
    case 1: // pending
      return (
        <div style={{ width: 200 }}>
          <Button
            variant="contained"
            color="secondary"
            className={classes.button}
            style={{
              backgroundColor: '#f44336',
            }}
            onClick={handleDecline}
            disabled={loading}
          >
            Decline
          </Button>
          <Button
            variant="contained"
            color="primary"
            className={classes.button}
            style={{
              backgroundColor: '#00cf6e',
            }}
            onClick={handleAccept}
            disabled={loading}
          >
            Accept
          </Button>
        </div>
      )
    case 2: // declined
      return (
        <Button
          variant="contained"
          className={classes.button}
          style={{
            backgroundColor: '#f44336',
          }}
          onClick={handleReset}
          disabled={loading}
        >
          Reset
        </Button>
      )
    case 4: // active
      return (
        <Button
          variant="contained"
          className={classes.button}
          style={{
            backgroundColor: '#00cf6e',
          }}
          onClick={handleAccept}
          disabled={loading}
        >
          Resend
        </Button>
      )
    default:
      return null
  }
}

const ControlPanelContainer = ({ data }) => {
  const classes = useStyles()
  const header = ['Email', 'Status', 'Action']
  const activeUsersCount = 0
  // statuses
  // 1 = new / pending
  // 2 = decline
  // 3 = resend
  // 4 = active
  const getStatusValue = (status) => {
    switch (Number(status)) {
      case 1:
        return 'Pending'
      case 2:
        return 'Declined'
      case 4:
        return 'Accepted'
      default:
        return ''
    }
  }
  // eslint-disable-next-line radix
  const inviteRequestCount = data.userInviteRequests.filter((user) => parseInt(user.status) === 1).length
  const totalUsers = data.userInviteRequests.length
  const result = data.userInviteRequests.reduce((r, { joined }) => {
    const dateObj = moment(joined).format('yyyy-MM-01')
    const objectKey = dateObj.toLocaleString('en-us', { year: 'numeric', month: 'numeric', day: 'numeric' })
    if (!r[objectKey]) r[objectKey] = { objectKey, entries: 1 }
    else r[objectKey].entries++
    return r
  }, {})
  const labels = Object.keys(result).sort((a, b) => new Date(a) - new Date(b))
  const lineSeries = labels.map((label) => result[label].entries)
  const formatLabels = labels.map((label) => {
    const dateObj = new Date(label)
    return dateObj.toLocaleString('en-us', { month: 'numeric', year: 'numeric' })
  })
  const chartData = {
    labels: formatLabels,
    series: [lineSeries],
  }
  const chartOptions = {
    lineSmooth: Chartist.Interpolation.cardinal({
      tension: 0,
    }),
    low: 0,
    high: totalUsers + 5,
    chartPadding: {
      top: 0,
      right: 0,
      bottom: 0,
      left: 0,
    },
  }
  return (
    <Grid container spacing={2} className={classes.panelContainer}>
      <Grid container>
        <Typography className={classes.panelHeader}>Invite Control Panel</Typography>
      </Grid>
      <Grid container spacing={4}>
        <Grid item xs={12} md={7} lg={7} xl={6}>
          <Card>
            <CardContent>
              <Typography className={classes.cardHeader}>User Invitation Requests</Typography>
              <TableContainer className={classes.tableContainer}>
                <Table className={classes.table} aria-label="simple table" stickyHeader>
                  <TableHead classes={{ head: classes.columnHeader }}>
                    <TableRow>
                      {header.map((name) => (
                        <TableCell
                          align="center"
                          className={classes.columnHeader}
                        >
                          {name}
                        </TableCell>
                      ))}
                    </TableRow>
                  </TableHead>
                  <TableBody>
                    {data.userInviteRequests.map((row) => (
                      <TableRow key={row._id}>
                        <TableCell align="left">{row.email}</TableCell>
                        <TableCell align="center">
                          <Button
                            variant="contained"
                            color="primary"
                            className={cx({
                              [classes.pendingStatus]: row.status === '1',
                              [classes.declinedStatus]: row.status === '2',
                              [classes.acceptedStatus]: row.status === '4',
                            })}
                            disableRipple
                            disableElevation
                          >
                            {getStatusValue(row.status)}
                          </Button>
                        </TableCell>
                        <TableCell align="center">
                          <ActionButtons status={row.status} id={row._id} />
                        </TableCell>
                      </TableRow>
                    ))}
                  </TableBody>
                </Table>
              </TableContainer>
            </CardContent>
          </Card>
        </Grid>
        <Grid item xs={12} md={5} lg={5} xl={6} justify="center">
          <Card>
            <CardContent>
              <Typography
                className={classes.cardHeader}
                display="inline"
              >
                User Invitation Statistics
              </Typography>
              <Typography
                className={classes.graphText}
                display="inline"
                style={{ float: 'right' }}
              >
                Invite Requests:
                {' '}
                {inviteRequestCount || 0}
              </Typography>
              <ChartistGraph
                className="ct-chart-white-colors"
                style={{ backgroundColor: '#00bcd4', marginTop: 10, marginBottom: 15 }}
                data={chartData}
                type="Line"
                options={chartOptions}
                listener={dailySalesChart.animation}
              />
              <Typography
                className={classes.graphText}
                display="inline"
              >
                Total Users:
                {' '}
                {totalUsers || 0}
              </Typography>
              <Typography
                className={classes.graphText}
                display="inline"
                style={{ float: 'right' }}
              >
                Active Users Today:
                {' '}
                {activeUsersCount}
              </Typography>
            </CardContent>
          </Card>
        </Grid>
      </Grid>
    </Grid>
  )
}

const ControlPanel = () => {
  const { data } = useQuery(USER_INVITE_REQUESTS)
  const classes = useStyles()
  if (data) {
    return (
      <ControlPanelContainer data={data} />
    )
  }
  return (
    <Grid container spacing={2} className={classes.panelContainer}>
      <Grid item xs={12}>
        <Skeleton animation="wave" style={{ width: '25%' }} />
      </Grid>
      <Grid container item xs={12}>
        <Grid container item xs={6} className={classes.sectionBorder}>
          <Skeleton animation="wave" height={300} style={{ width: '80%' }} />
        </Grid>
        <Grid container item xs={6} justify="flex-end">
          <Skeleton animation="wave" height={300} style={{ width: '80%' }} />
        </Grid>
      </Grid>
    </Grid>
  )
}

ControlPanelContainer.propTypes = {
  data: PropTypes.object.isRequired,
}

ActionButtons.propTypes = {
  status: PropTypes.string.isRequired,
  id: PropTypes.string.isRequired,
}

export default ControlPanel
