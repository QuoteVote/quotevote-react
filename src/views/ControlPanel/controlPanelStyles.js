const requestAccessStyles = (theme) => ({
  tableContainer: {
    maxHeight: 500,
  },
  panelContainer: {
    [theme.breakpoints.up('lg')]: {
      padding: 20,
      paddingTop: 30,
    },
    [theme.breakpoints.down('md')]: {
      padding: 10,
    },
    [theme.breakpoints.down('xs')]: {
      padding: 0,
    },
  },
  panelHeader: {
    font: 'Montserrat',
    fontSize: '24px',
    fontWeight: 'bold',
    color: 'black',
    marginBottom: 25,
  },
  cardHeader: {
    font: 'Montserrat',
    fontSize: '18px',
    fontWeight: 'bold',
    letterSpacing: '0.2px',
    color: '#00cf6e',
    marginBottom: 10,
  },
  columnHeader: {
    font: 'Roboto',
    fontSize: '17px',
  },
  button: {
    width: '83.1px',
    height: '20.8px',
    borderRadius: '3px',
    textTransform: 'none',
    color: 'white',
    margin: 5,
  },
  pendingStatus: {
    borderRadius: '10px',
    backgroundColor: '#d8d8d8',
    '&:hover': {
      backgroundColor: '#d8d8d8',
    },
    width: '83.1px',
    height: '20.8px',
    textTransform: 'none',
    margin: 5,
    color: 'black',
    cursor: 'default',
  },
  acceptedStatus: {
    borderRadius: '10px',
    backgroundColor: '#4caf50',
    '&:hover': {
      backgroundColor: '#4caf50',
    },
    width: '83.1px',
    height: '20.8px',
    textTransform: 'none',
    margin: 5,
    color: 'white',
    cursor: 'default',
  },
  declinedStatus: {
    borderRadius: '10px',
    backgroundColor: '#ff6060',
    '&:hover': {
      backgroundColor: '#ff6060',
    },
    width: '83.1px',
    height: '20.8px',
    textTransform: 'none',
    margin: 5,
    color: 'white',
    cursor: 'default',
  },
  graphText: {
    font: 'Roboto',
    fontSize: '18px',
    fontWeight: 300,
    lineHeight: 1.39,
    color: '#333333',
  },
  featuredRow: {
    backgroundColor: '#fff8e1',
  },
  slotSelect: {
    minWidth: 80,
  },
  filterInput: {
    marginBottom: 15,
  },
})

export default requestAccessStyles
