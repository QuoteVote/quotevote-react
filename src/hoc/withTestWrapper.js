import React from 'react'
import { PersistGate } from 'redux-persist/integration/react'
import { ApolloProvider } from '@apollo/react-hooks'
import { Provider } from 'react-redux'
import { createTheme, ThemeProvider } from '@material-ui/core/styles'
import { Router } from 'react-router-dom'
import { createBrowserHistory } from 'history'
import store, { persistor } from '../store/store'
import client from '../config/apollo'
import customTheme from '../theme'
const hist = createBrowserHistory()
const theme = createTheme(customTheme)

const withTestWrapper = (Component) => (props) => (
  <ApolloProvider client={client}>
    <Provider store={store}>
      <PersistGate loading={null} persistor={persistor}>
        <ThemeProvider theme={theme}>
          <Router history={hist}>
            <Component {...props} />
          </Router>
        </ThemeProvider>
      </PersistGate>
    </Provider>
  </ApolloProvider>
)
export default withTestWrapper
