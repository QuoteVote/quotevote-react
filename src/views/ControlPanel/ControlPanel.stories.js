// Important stuff, must always be imported on a storybook file
import React from 'react'
import { withKnobs } from '@storybook/addon-knobs/react'
import { withA11y } from '@storybook/addon-a11y'

// If you want to apply theme
import { MuiThemeProvider as ThemeProvider } from '@material-ui/core/styles'

// The component
import ControlPanel from './ControlPanel'

import theme from '../../themes/MainTheme'

// Story config
export default {
  title: 'ControlPanel',
  argTypes: { onClick: { action: 'clicked' }, onChange: { action: 'typing' } },
  component: ControlPanel,
  decorators: [withKnobs, withA11y],
}

function Wrapper() {
  return (
    <ThemeProvider theme={theme}>
      <ControlPanel />
    </ThemeProvider>
  )
}

export const base = () => <Wrapper />

base.story = {
  parameters: {
    jest: ['ControlPanel.test.js'],
  },
}
