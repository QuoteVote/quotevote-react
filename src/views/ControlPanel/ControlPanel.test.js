import React from 'react'
import { render } from '@testing-library/react'

// Component being tested
import ControlPanel from './ControlPanel'
import withTestWrapper from '../../hoc/withTestWrapper'

const ControlPanelWrapper = withTestWrapper(ControlPanel)

describe('ControlPanel test -', () => {
  it('renders correctly', () => {
    const { container } = render(
      <ControlPanelWrapper />,
    )
    expect(container.firstChild).toMatchSnapshot()
  })
})
