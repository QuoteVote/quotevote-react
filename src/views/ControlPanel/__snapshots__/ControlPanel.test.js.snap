// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ControlPanel test - renders correctly 1`] = `null`;
